source/adapters/hip/image.cpp

```diff
@@ -8,178 +8,1335 @@
+urToHipImageChannelFormat(ur_image_channel_type_t image_channel_type,
+                          ur_image_channel_order_t image_channel_order,
+                          hipArray_Format *return_hip_format,
+                          size_t *return_pixel_size_bytes,
+                          unsigned int *return_normalized_dtype_flag) {
+
+  hipArray_Format hip_format = HIP_AD_FORMAT_UNSIGNED_INT8;
+  size_t pixel_size_bytes = 0;
+  unsigned int num_channels = 0;
+  unsigned int normalized_dtype_flag = 0;
+  UR_CHECK_ERROR(urCalculateNumChannels(image_channel_order, &num_channels));
+
+  switch (image_channel_type) {
+#define CASE(FROM, TO, SIZE, NORM)                                             \
+  case FROM: {                                                                 \
+    hip_format = TO;                                                           \
+    pixel_size_bytes = SIZE * num_channels;                                    \
+    normalized_dtype_flag = NORM;                                              \
+    break;                                                                     \
+  }
+
+    CASE(UR_IMAGE_CHANNEL_TYPE_UNSIGNED_INT8, HIP_AD_FORMAT_UNSIGNED_INT8, 1, 0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_SIGNED_INT8, HIP_AD_FORMAT_SIGNED_INT8, 1, 0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_UNSIGNED_INT16, HIP_AD_FORMAT_UNSIGNED_INT16, 2,
+         0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_SIGNED_INT16, HIP_AD_FORMAT_SIGNED_INT16, 2, 0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_HALF_FLOAT, HIP_AD_FORMAT_HALF, 2, 0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_UNSIGNED_INT32, HIP_AD_FORMAT_UNSIGNED_INT32, 4,
+         0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_SIGNED_INT32, HIP_AD_FORMAT_SIGNED_INT32, 4, 0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_FLOAT, HIP_AD_FORMAT_FLOAT, 4, 0)
+    CASE(UR_IMAGE_CHANNEL_TYPE_UNORM_INT8, HIP_AD_FORMAT_UNSIGNED_INT8, 1, 1)
+    CASE(UR_IMAGE_CHANNEL_TYPE_SNORM_INT8, HIP_AD_FORMAT_SIGNED_INT8, 1, 1)
+    CASE(UR_IMAGE_CHANNEL_TYPE_UNORM_INT16, HIP_AD_FORMAT_UNSIGNED_INT16, 2, 1)
+    CASE(UR_IMAGE_CHANNEL_TYPE_SNORM_INT16, HIP_AD_FORMAT_SIGNED_INT16, 2, 1)
+
+#undef CASE
+  default:
+    break;
+  }
+
+  if (return_hip_format) {
+    *return_hip_format = hip_format;
+  }
+  if (return_pixel_size_bytes) {
+    *return_pixel_size_bytes = pixel_size_bytes;
+  }
+  if (return_normalized_dtype_flag) {
+    *return_normalized_dtype_flag = normalized_dtype_flag;
+  }
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t
+hipToUrImageChannelFormat(hipArray_Format hip_format,
+                          ur_image_channel_type_t *return_image_channel_type) {
+
+  switch (hip_format) {
+#define HIP_TO_UR_IMAGE_CHANNEL_TYPE(FROM, TO)                                 \
+  case FROM: {                                                                 \
+    *return_image_channel_type = TO;                                           \
+    return UR_RESULT_SUCCESS;                                                  \
+  }
+    HIP_TO_UR_IMAGE_CHANNEL_TYPE(HIP_AD_FORMAT_UNSIGNED_INT8,
```

Comment:
I note that this could also map to `UR_IMAGE_CHANNEL_TYPE_UNORM_INT8` (going by `urToHipImageChannelFormat` above). Is it a problem that we can't round-trip these? Does the UR spec expect a unique mapping? Should we document this somewhere?